cpp/src/arrow/compare.cc

```diff
@@ -1207,12 +1255,13 @@ inline bool SparseTensorEqualsImplDispatch(const SparseTensorImpl<SparseIndexTyp
 
 }  // namespace
 
-bool SparseTensorEquals(const SparseTensor& left, const SparseTensor& right) {
-  if (&left == &right) {
+bool SparseTensorEquals(const SparseTensor& left, const SparseTensor& right,
+                        const EqualOptions& opts) {
+  if (is_integer(left.type()->id()) && &left == &right) {
```

Comment:
You can probably move this shortcut into `IntegerSparseTensorDataEquals` and `FloatingSparseTensorDataEquals` (so that you can also enable the shortcut for FP values if `nans_equal` is true).

Reply:
Sounds reasonable!